File: 3355-zero-array-transformation-i/3355-zero-array-transformation-i.cpp
class Solution {
public:
    bool isZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        vector<int> prefix(n+1);
        for(auto q:queries) {
            prefix[q[0]]++;
            prefix[q[1]+1]--;
        }

        for(int i=1;i<=n;i++){
            prefix[i]+=prefix[i-1];
            if(prefix[i-1]<nums[i-1]){
                return false;
            }
        }
        return true;
    }
};


/*

Query = [2,5]

0   1   2   3   4   5   6   
2   5   9   4   1   0   3

0   0   0   0   0   0   0
0   0   1   0   0   0   -1
0   0   1   1   1   1   0




*/

File: 3355-zero-array-transformation-i/3355-zero-array-transformation-i.java
class Solution {
    public boolean isZeroArray(int[] nums, int[][] queries) {
        int n = nums.length;
        int[] prefix = new int[n+1];
        for(int i=0;i<queries.length;i++) {
            int lowerIndex = queries[i][0];
            int upperIndex = queries[i][1]+1;
            prefix[lowerIndex]--;
            prefix[upperIndex]++;
        }

        for(int i=1;i<prefix.length;i++) {
            if(prefix[i-1]+nums[i-1]>0) return false;
            prefix[i]=prefix[i-1]+prefix[i];
        }

        return true;
    }
}

/*
0   1   2   3
4   3   2   1



1   3
0   2


prefix
0   1   2   3   4
0   0   0   0   0
    -1          +1
-1          +1    

-1  -1  0   1   1


-1  -2  -2  -1  0

3   1   0   0


*/